command/agent/local.go

```diff
@@ -174,10 +174,14 @@ func (l *localState) RemoveService(serviceID string) {
 	l.Lock()
 	defer l.Unlock()
 
-	delete(l.services, serviceID)
-	delete(l.serviceTokens, serviceID)
-	l.serviceStatus[serviceID] = syncStatus{inSync: false}
-	l.changeMade()
+	if _, ok := l.services[serviceID]; ok {
+		delete(l.services, serviceID)
+		delete(l.serviceTokens, serviceID)
+		l.serviceStatus[serviceID] = syncStatus{inSync: false}
+		l.changeMade()
+	} else {
+		l.logger.Printf("[WARN] agent: Tried to deregister nonexistent service '%s'", serviceID)
```

Comment:
I think you'll want to add an error return to this function can compose the warning one level up. Otherwise you'll get this log message followed by "agent: removed service" which'll be confusing. I think up there we can return nil from `RemoveService()` and add a comment about how we are trying to keep the current idempotent behavior for now (probably a TODO to fix for 0.8).